components/user-store/org.wso2.carbon.identity.user.store.count/src/main/java/org/wso2/carbon/identity/user/store/count/UserStoreCountService.java

```diff
@@ -181,27 +163,87 @@ public Long countRolesInDomain(String filter, String domain) throws UserStoreCou
-            UserStoreCounterException {
-
-        UserStoreCountRetriever counter = UserStoreCountUtils.getCounterInstanceForDomain(domain);
-        if (counter != null) {
-            return counter.countClaim(claimURI, valueFilter);
-        } else {
-            return Long.valueOf(-1);
-        }
+    public long countByClaimInDomain(String claimURI, String valueFilter, String domain)
+            throws UserStoreCounterException {
 
+        String filterWithDomain = getFilterWithDomain(domain, valueFilter);
+        return getUserCountWithClaims(claimURI, filterWithDomain);
     }
 
     /**
      * Get count enabled user stores.
+     *
      * @return
      * @throws UserStoreCounterException
      */
     public String[] getCountEnabledUserStores() throws UserStoreCounterException {
+
         Set<String> domains = UserStoreCountUtils.getCountEnabledUserStores();
         return domains.toArray(new String[domains.size()]);
 
     }
 
+    /**
+     * Get User count.
+     *
+     * @param claimURI    claim uri
+     * @param valueFilter filter
+     * @return user count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getUserCountWithClaims(String claimURI, String valueFilter) throws UserStoreCounterException {
+
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            UserStoreManager userStoreManager =
+                    UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+            if (userStoreManager instanceof org.wso2.carbon.user.core.UserStoreManager) {
+                return ((org.wso2.carbon.user.core.UserStoreManager) userStoreManager).getUserCountWithClaims(claimURI,
+                        valueFilter);
+            } else {
+                throw new UserStoreCounterException("Error while retrieving User Store from core");
+            }
+
+        } catch (UserStoreException e) {
+            String errorMsg = "error while retrieving user count from user store";
+            throw new UserStoreCounterException(errorMsg, e);
+        }
+    }
+
+    /**
+     * Get role count.
+     *
+     * @param filter filter
+     * @return role count
+     * @throws UserStoreCounterException UserStoreCounterException
+     */
+    private long getCountRoles(String filter) throws UserStoreCounterException {
+
+        try {
+            realmConfiguration = CarbonContext.getThreadLocalCarbonContext().getUserRealm().getRealmConfiguration();
+            UserStoreManager userStoreManager =
+                    UserStoreCountDSComponent.getRealmService().getTenantUserRealm(tenantId).getUserStoreManager();
+            if (userStoreManager instanceof org.wso2.carbon.user.core.UserStoreManager) {
+                return ((org.wso2.carbon.user.core.UserStoreManager) userStoreManager).countRoles(filter);
```

Comment:
Does the `filter` is domain qualified? better to include that in the java doc comment.